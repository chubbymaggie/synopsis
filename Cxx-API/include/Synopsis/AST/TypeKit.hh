//
// Copyright (C) 2004 Stefan Seefeld
// All rights reserved.
// Licensed to the public under the terms of the GNU LGPL (>= 2),
// see the file COPYING for details.
//

#ifndef _Synopsis_AST_TypeKit_hh
#define _Synopsis_AST_TypeKit_hh

#include <Synopsis/Python/Kit.hh>
#include <Synopsis/AST/Type.hh>
#include <Synopsis/AST/Declared.hh>

namespace Synopsis
{
namespace AST
{

// basically a factory for all Type types
class TypeKit : public Python::Kit
{
public:
  TypeKit(std::string const &lang) : Kit("Synopsis.Type"), my_lang(lang) {}

  Type create_type()
  { return create<Type>("Type", Python::Tuple(my_lang));}

  Named create_named(const ScopedName &sn)
  { return create<Named>("Named", Python::Tuple(my_lang, sn));}

  Base create_base(const ScopedName &sn)
  { return create<Base>("Base", Python::Tuple(my_lang, sn));}

  Dependent create_dependent(const ScopedName &sn)
  { return create<Dependent>("Dependent", Python::Tuple(my_lang, sn));}

  Unknown create_unknown(const ScopedName &sn)
  { return create<Unknown>("Unknown", Python::Tuple(my_lang, sn));}

  Declared create_declared(const ScopedName &sn,
			   const Declaration &decl)
  { return create<Declared>("Declared", Python::Tuple(my_lang, sn, decl));}

  Template create_template(const ScopedName &sn,
			   const Declaration &decl, const Python::List &params)
  { return create<Template>("Template", Python::Tuple(my_lang, sn, decl, params));}

  Modifier create_modifier(const Type &alias,
			   const Modifiers &pre, const Modifiers &post)
  { return create<Modifier>("Modifier", Python::Tuple(my_lang, alias, pre, post));}

  Array create_array(const Type &alias,
		     const Python::TypedList<int> &sizes)
  { return create<Array>("Array", Python::Tuple(my_lang, alias, sizes));}

  Parametrized create_parametrized(const Template &t,
				   const Python::List &params)
  { return create<Parametrized>("Parametrized", Python::Tuple(my_lang, t, params));}

  FunctionPtr create_function_ptr(const Type &retn,
				  const Modifiers &pre, const TypeList &params)
  { return create<FunctionPtr>("Function", Python::Tuple(my_lang, retn, pre, params));}

private:
  std::string my_lang;
};

}
}

#endif
