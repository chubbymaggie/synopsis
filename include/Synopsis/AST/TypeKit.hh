// $Id: TypeKit.hh,v 1.1 2004/01/25 21:21:54 stefan Exp $
//
// Copyright (C) 2004 Stefan Seefeld
// All rights reserved.
// Licensed to the public under the terms of the GNU LGPL (>= 2),
// see the file COPYING for details.
//

#ifndef _Synopsis_AST_TypeKit_hh
#define _Synopsis_AST_TypeKit_hh

#include <Synopsis/Kit.hh>
#include <Synopsis/AST/Type.hh>
#include <Synopsis/AST/Declared.hh>

namespace Synopsis
{
namespace AST
{

// basically a factory for all Type types
class TypeKit : public Kit
{
public:
  TypeKit() : Kit("Synopsis.Type") {}

  Type create_type(const std::string &lang)
  { return create<Type>("Type", Tuple(lang));}

  Named create_named(const std::string &lang, const ScopedName &sn)
  { return create<Named>("Named", Tuple(lang, sn));}

  Base create_base(const std::string &lang, const ScopedName &sn)
  { return create<Base>("Base", Tuple(lang, sn));}

  Dependent create_dependent(const std::string &lang, const ScopedName &sn)
  { return create<Dependent>("Dependent", Tuple(lang, sn));}

  Unknown create_unknown(const std::string &lang, const ScopedName &sn)
  { return create<Unknown>("Unknown", Tuple(lang, sn));}

  Declared create_declared(const std::string &lang, const ScopedName &sn,
			   const Declaration &decl)
  { return create<Declared>("Declared", Tuple(lang, sn, decl));}

  Template create_template(const std::string &lang, const ScopedName &sn,
			   const Declaration &decl, const List &params)
  { return create<Template>("Template", Tuple(lang, sn, decl, params));}

  Modifier create_modifier(const std::string &lang, const Type &alias,
			   const std::string &pre, const std::string &post)
  { return create<Modifier>("Modifier", Tuple(lang, alias, pre, post));}

  Array create_array(const std::string &lang, const Type &alias,
		     const TypedList<size_t> &sizes)
  { return create<Array>("Array", Tuple(lang, alias, sizes));}

  Parametrized create_parametrized(const std::string &lang, const Template &t,
				   const List &params)
  { return create<Parametrized>("Parametrized", Tuple(lang, t, params));}

  FunctionPtr create_function_ptr(const std::string &lang, const Type &retn,
				  const Modifiers &pre, const TypeList &params)
  { return create<FunctionPtr>("Function", Tuple(lang, retn, pre, params));}

};

}
}

#endif
