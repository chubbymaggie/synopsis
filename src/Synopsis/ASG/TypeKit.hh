//
// Copyright (C) 2004 Stefan Seefeld
// All rights reserved.
// Licensed to the public under the terms of the GNU LGPL (>= 2),
// see the file COPYING for details.
//

#ifndef _Synopsis_ASG_TypeKit_hh
#define _Synopsis_ASG_TypeKit_hh

#include <Synopsis/Python/Kit.hh>
#include <Synopsis/ASG/Type.hh>
#include <Synopsis/ASG/Declared.hh>

namespace Synopsis
{
namespace ASG
{

// basically a factory for all Type types
class TypeKit : public Python::Kit
{
public:
  TypeKit(std::string const &lang) : Kit("Synopsis.Type"), lang_(lang) {}

  Type create_type()
  { return create<Type>("Type", Python::Tuple(lang_));}

  Named create_named(const ScopedName &sn)
  { return create<Named>("Named", Python::Tuple(lang_, sn));}

  Base create_base(const ScopedName &sn)
  { return create<Base>("Base", Python::Tuple(lang_, sn));}

  Dependent create_dependent(const ScopedName &sn)
  { return create<Dependent>("Dependent", Python::Tuple(lang_, sn));}

  Unknown create_unknown(const ScopedName &sn)
  { return create<Unknown>("Unknown", Python::Tuple(lang_, sn));}

  Declared create_declared(const ScopedName &sn,
			   const Declaration &decl)
  { return create<Declared>("Declared", Python::Tuple(lang_, sn, decl));}

  Template create_template(const ScopedName &sn,
			   const Declaration &decl, const Python::List &params)
  { return create<Template>("Template", Python::Tuple(lang_, sn, decl, params));}

  Modifier create_modifier(const Type &alias,
			   const Modifiers &pre, const Modifiers &post)
  { return create<Modifier>("Modifier", Python::Tuple(lang_, alias, pre, post));}

  Array create_array(const Type &alias,
		     const Python::TypedList<int> &sizes)
  { return create<Array>("Array", Python::Tuple(lang_, alias, sizes));}

  Parametrized create_parametrized(const Template &t,
				   const Python::List &params)
  { return create<Parametrized>("Parametrized", Python::Tuple(lang_, t, params));}

  FunctionPtr create_function_ptr(const Type &retn,
				  const Modifiers &pre, const TypeList &params)
  { return create<FunctionPtr>("Function", Python::Tuple(lang_, retn, pre, params));}

private:
  std::string lang_;
};

}
}

#endif
