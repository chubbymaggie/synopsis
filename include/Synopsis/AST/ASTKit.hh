//
// Copyright (C) 2004 Stefan Seefeld
// All rights reserved.
// Licensed to the public under the terms of the GNU LGPL (>= 2),
// see the file COPYING for details.
//

#ifndef _Synopsis_AST_ASTKit_hh
#define _Synopsis_AST_ASTKit_hh

#include <Synopsis/Kit.hh>
#include <Synopsis/AST/AST.hh>
#include <Synopsis/AST/SourceFile.hh>
#include <Synopsis/AST/Declaration.hh>

namespace Synopsis
{
namespace AST
{

// basically a factory for all AST types
class ASTKit : public Kit
{
public:
  ASTKit() : Kit("Synopsis.AST") {}

  AST create_ast() { return create<AST>("AST");}

  Comment create_comment(const SourceFile &file, long line,
			 const std::string &text, bool suspect=false)
  { return create<Comment>("Declaration", Tuple(text, file, line, suspect));}
  
  Declaration create_declaration(const SourceFile &sf, long line, const std::string &lang,
				 const char *type, const ScopedName &name)
  { return create<Declaration>("Declaration", Tuple(sf, line, lang, type, name));}

  Builtin create_builtin(const SourceFile &file, int line, const std::string &lang,
			 const std::string &type, const ScopedName &name)
  { return create<Builtin>("Builtin", Tuple(file, line, lang, type, name));}

  Include create_include(const SourceFile &sf, bool is_macro, bool is_next)
  { return create<Include>("Include", Tuple(sf, is_macro, is_next));}

  Macro create_macro(SourceFile &sf, long line, const std::string &lang,
		     const ScopedName &name, const List &parameters,
		     const std::string &text)
  { return create<Macro>("Macro", Tuple(sf, line, lang, "macro",
					name, parameters, text));}

  Forward create_forward(const SourceFile &file, int line, const std::string &lang,
			 const std::string &type, const ScopedName &name)
  { return create<Forward>("Forward", Tuple(file, line, lang, type, name));}

  MacroCall create_macro_call(const std::string &name, int start, int end, int diff)
  { return create<MacroCall>("MacroCall", Tuple(name, start, end, diff));}

  Scope create_scope(const SourceFile &file, int line, const std::string &lang,
		     const std::string &type, const ScopedName &name)
  { return create<Scope>("Scope", Tuple(file, line, lang, type, name));}

  Synopsis::AST::Module
  create_module(const SourceFile &file, int line, const std::string &lang,
		const std::string &type, const ScopedName &name)
  { return create<Synopsis::AST::Module>("Module", Tuple(file, line, lang, type, name));}

  Inheritance create_inheritance(const Type &parent,
				 const List &attributes)
  { return create<Inheritance>("Inheritance", Tuple(parent, attributes));}

  Class create_class(const SourceFile &file, int line, const std::string &lang,
		     const std::string &type, const ScopedName &name)
  { return create<Class>("Class", Tuple(file, line, lang, type, name));}

  Typedef create_typedef(const SourceFile &file, int line, const std::string &lang,
			 const std::string &type, const ScopedName &name,
			 const Type &alias, bool constr)
  { return create<Typedef>("Typedef", Tuple(file, line, lang, type, name, alias, constr));}

  Enumerator create_enumerator(const SourceFile &file, int line, const std::string &lang,
			       const std::string &type, const ScopedName &name,
			       const std::string &value)
  { return create<Enumerator>("Enumerator", Tuple(file, line, lang, type, name, value));}

  Enum create_enum(const SourceFile &file, int line, const std::string &lang,
		   const std::string &type, const ScopedName &name)
  { return create<Enum>("Enum", Tuple(file, line, lang, type, name));}

  Variable create_variable(const SourceFile &file, int line, const std::string &lang,
			   const std::string &type, const ScopedName &name,
			   const Type &vtype, bool constr)
  { return create<Variable>("Variable", Tuple(file, line, lang, type, name, vtype, constr));}

  Const create_const(const SourceFile &file, int line, const std::string &lang,
		     const std::string &type, const ScopedName &name,
		     const Type &ctype, const std::string &value)
  { return create<Const>("Const", Tuple(file, line, lang, type, name, ctype, value));}

  Parameter create_parameter(const Modifiers &pre, const Type &type, const Modifiers &post,
			     const std::string &name, const std::string &value)
  { return create<Parameter>("Parameter", Tuple(pre, type, post, name, value));}

  Function create_function(const SourceFile &file, int line, const std::string &lang,
 			   const std::string &type, const Modifiers &pre,
 			   const Type &ret, const ScopedName &name,
 			   const std::string &realname)
  { return create<Function>("Function", Tuple(file, line, lang, type, pre, ret,
 					      name, realname));}

  Operation create_operation(const SourceFile &file, int line, const std::string &lang,
			     const std::string &type, const ScopedName &name,
			     const Modifiers &pre, const Type &ret,
			     const std::string &realname)
  { return create<Operation>("Operation", Tuple(file, line, lang, type, name,
						pre, ret, realname));}

  SourceFile create_source_file(const std::string &name,
				const std::string &longname,
				const std::string &lang)
  { return create<SourceFile>("SourceFile", Tuple(name, longname, lang));}

};

}
}

#endif
