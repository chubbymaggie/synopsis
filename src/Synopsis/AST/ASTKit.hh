//
// Copyright (C) 2004 Stefan Seefeld
// All rights reserved.
// Licensed to the public under the terms of the GNU LGPL (>= 2),
// see the file COPYING for details.
//

#ifndef _Synopsis_AST_ASTKit_hh
#define _Synopsis_AST_ASTKit_hh

#include <Synopsis/Python/Kit.hh>
#include <Synopsis/AST/AST.hh>
#include <Synopsis/AST/SourceFile.hh>
#include <Synopsis/AST/Declaration.hh>

namespace Synopsis
{
namespace AST
{

// basically a factory for all AST types
class ASTKit : public Python::Kit
{
public:
  ASTKit(std::string const &lang) : Python::Kit("Synopsis.AST"), my_lang(lang) {}

  AST create_ast() { return create<AST>("AST");}

  Comment create_comment(const SourceFile &file, long line,
			 const std::string &text, bool suspect=false)
  { return create<Comment>("Comment", Python::Tuple(text, file, line, suspect));}
  
  Declaration create_declaration(const SourceFile &sf, long line,
				 const char *type, const ScopedName &name)
  { return create<Declaration>("Declaration", Python::Tuple(sf, line, my_lang, type, name));}

  Builtin create_builtin(const SourceFile &file, int line,
			 const std::string &type, const ScopedName &name)
  { return create<Builtin>("Builtin", Python::Tuple(file, line, my_lang, type, name));}

  Include create_include(const SourceFile &sf, const std::string &name,
			 bool is_macro, bool is_next)
  { return create<Include>("Include", Python::Tuple(sf, name, is_macro, is_next));}

  Macro create_macro(SourceFile &sf, long line,
		     const ScopedName &name, const Python::List &parameters,
		     const std::string &text)
  { return create<Macro>("Macro", Python::Tuple(sf, line, my_lang, "macro",
					name, parameters, text));}

  Forward create_forward(const SourceFile &file, int line,
			 const std::string &type, const ScopedName &name)
  { return create<Forward>("Forward", Python::Tuple(file, line, my_lang, type, name));}

  MacroCall create_macro_call(const std::string &name, int start, int end, int diff)
  { return create<MacroCall>("MacroCall", Python::Tuple(name, start, end, diff));}

  Scope create_scope(const SourceFile &file, int line,
		     const std::string &type, const ScopedName &name)
  { return create<Scope>("Scope", Python::Tuple(file, line, my_lang, type, name));}

  Synopsis::AST::Module
  create_module(const SourceFile &file, int line,
		const std::string &type, const ScopedName &name)
  { return create<Synopsis::AST::Module>("Module", Python::Tuple(file, line, my_lang, type, name));}

  Inheritance create_inheritance(const Type &parent,
				 const Python::List &attributes)
  { return create<Inheritance>("Inheritance", Python::Tuple(parent, attributes));}

  Class create_class(const SourceFile &file, int line,
		     const std::string &type, const ScopedName &name)
  { return create<Class>("Class", Python::Tuple(file, line, my_lang, type, name));}

  Typedef create_typedef(const SourceFile &file, int line,
			 const std::string &type, const ScopedName &name,
			 const Type &alias, bool constr)
  { return create<Typedef>("Typedef", Python::Tuple(file, line, my_lang, type, name, alias, constr));}

  Enumerator create_enumerator(const SourceFile &file, int line,
			       const ScopedName &name, const std::string &value)
  { return create<Enumerator>("Enumerator", Python::Tuple(file, line, my_lang, name, value));}

  Enum create_enum(const SourceFile &file, int line,
		   const ScopedName &name, const Enumerators &values)
  { return create<Enum>("Enum", Python::Tuple(file, line, my_lang, name, values));}

  Variable create_variable(const SourceFile &file, int line,
			   const std::string &type, const ScopedName &name,
			   const Type &vtype, bool constr)
  { return create<Variable>("Variable", Python::Tuple(file, line, my_lang, type, name, vtype, constr));}

  Const create_const(const SourceFile &file, int line,
		     const std::string &type, const ScopedName &name,
		     const Type &ctype, const std::string &value)
  { return create<Const>("Const", Python::Tuple(file, line, my_lang, type, name, ctype, value));}

  Parameter create_parameter(const Modifiers &pre, const Type &type, const Modifiers &post,
			     const std::string &name, const std::string &value)
  { return create<Parameter>("Parameter", Python::Tuple(pre, type, post, name, value));}

  Function create_function(const SourceFile &file, int line,
 			   const std::string &type, const Modifiers &pre,
 			   const Type &ret, const Modifiers &post,
			   const ScopedName &name, const std::string &realname)
  { return create<Function>("Function", Python::Tuple(file, line, my_lang, type, pre, ret, post,
						      name, realname));}

  Operation create_operation(const SourceFile &file, int line,
			     const std::string &type, const Modifiers &pre,
			     const Type &ret, const Modifiers &post,
			     const ScopedName &name, const std::string &realname)
  { return create<Operation>("Operation", Python::Tuple(file, line, my_lang, type, pre, ret, post,
							name, realname));}

  SourceFile create_source_file(const std::string &name,
				const std::string &longname)
  { return create<SourceFile>("SourceFile", Python::Tuple(name, longname, my_lang));}
private:
  std::string my_lang;
};

}
}

#endif
