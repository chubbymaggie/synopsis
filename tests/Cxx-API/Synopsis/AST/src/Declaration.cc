#include <Synopsis/AST/ASTKit.hh>
#include <Synopsis/AST/TypeKit.hh>
#include "Guard.hh"
#include <string>
#include <iostream>

using namespace Synopsis;

void test1()
{
  AST::ASTKit kit = AST::ASTKit();
  AST::TypeKit types = AST::TypeKit();
  AST::SourceFile sf = kit.create_source_file("filename", "/long/filename", "C++");
  AST::Declaration d = kit.create_declaration(sf, 2, "C--", "foo", Tuple("bar"));
  std::cout << d.attr("__class__") << std::endl;
  std::cout << "file='" << d.file().attr("__class__") << "\'\n"
	    << "line='" << d.line() << "\'\n"
	    << "language='" << d.language() << "\'\n"
	    << "type='" << d.type() << "\'\n"
	    << "name='" << d.name() << "\'\n"
	    << "comments='" << d.comments() << '\'' << std::endl;
  std::cout << kit.create_builtin(sf, 3, "C--", "eos", Tuple("eos")).attr("__class__") << std::endl;
  std::cout << kit.create_macro(sf, 4, "C--", Tuple("ZAP"), List(), "").attr("__class__") << std::endl;
  std::cout << kit.create_forward(sf, 5, "C--", "forward", Tuple("flip")).attr("__class__") << std::endl;
  std::cout << kit.create_module(sf, 6, "C--", "namespace", Tuple("flop")).attr("__class__") << std::endl;
  AST::Class c = kit.create_class(sf, 7, "C--", "class", Tuple("faz"));
  std::cout << c.attr("__class__") << std::endl;
  AST::Type t = types.create_declared("C--", Tuple("faz"), c);
  std::cout << kit.create_typedef(sf, 7, "C--", "class", Tuple("faz"), t, false).attr("__class__") << std::endl;
}

int main(int, char **)
{
  try
  {
    test1();
  }
  catch (const std::exception &e)
  {
    std::cout << "Error : " << e.what() << std::endl;
  }
}

