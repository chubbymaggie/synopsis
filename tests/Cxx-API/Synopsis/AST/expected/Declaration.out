Synopsis.AST.Declaration
file='Synopsis.AST.SourceFile'
line='2'
language='C--'
type='foo'
name='['bar']'
comments='[]'
Synopsis.AST.Builtin
Synopsis.AST.Macro
Synopsis.AST.Forward
Synopsis.AST.Module
Synopsis.AST.Class
Synopsis.AST.Typedef
